crates/runix/src/flake_ref/file.rs

```diff
@@ -0,0 +1,315 @@
+            url = self.url
+        ))
+        .unwrap();
+
+        url.set_query(
+            serde_urlencoded::to_string(&self.attributes)
+                .ok()
+                .filter(|s| !s.is_empty())
+                .as_deref(),
+        );
+
+        write!(f, "{url}")
+    }
+}
+
+impl<Protocol: FileProtocol, App: ApplicationProtocol> FromStr for FileBasedRef<Protocol, App> {
+    type Err = ParseFileError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let url = Url::parse(s)?;
+
+        // fix url
+        let (app, proto, url, wrapped) = if let Some((app, proto)) = &url.scheme().split_once('+') {
+            let wrapped = FileUrl::<Protocol>::from_str(s.trim_start_matches(&format!("{app}+")))?;
+
+            (app.to_string(), proto.to_string(), url, wrapped)
+        } else if App::required(&url) {
+            return Err(ParseFileError::InvalidScheme(
+                Self::scheme().to_string(),
+                url.scheme().to_string(),
+            ));
+        } else {
+            let app = App::protocol();
+            let proto = url.scheme();
+            let fixed = Url::parse(&format!("{app}+{url}"))?;
+            (
+                app.to_string(),
+                proto.to_string(),
+                fixed,
+                FileUrl::<Protocol>::from_str(s)?,
+            )
+        };
+
+        if app != App::protocol() || proto != Protocol::scheme() {
+            return Err(ParseFileError::InvalidScheme(
+                Self::scheme().to_string(),
+                url.scheme().to_string(),
+            ));
+        };
+
+        let attributes: FileAttributes =
+            serde_urlencoded::from_str(url.query().unwrap_or_default())?;
+
+        Ok(FileBasedRef {
+            url: wrapped,
+            attributes,
+            _type: Application::default(),
+        })
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum ParseFileError {
+    #[error(transparent)]
+    Url(#[from] url::ParseError),
+    #[error(transparent)]
+    FileUrl(#[from] WrappedUrlParseError),
+    #[error("Invalid scheme (expected: '{0}:', found '{1}:')")]
+    InvalidScheme(String, String),
+    #[error("No repo specified")]
+    NoRepo,
+    #[error("Couldn't parse query: {0}")]
+    Query(#[from] serde_urlencoded::de::Error),
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use crate::flake_ref::tests::{roundtrip, roundtrip_to};
+
+    type FileFileRef = super::FileRef<protocol::File>;
+    type HttpFileRef = super::FileRef<protocol::HTTP>;
+    type HttpsFileRef = super::FileRef<protocol::HTTPS>;
+
+    type FileTarballRef = super::TarballRef<protocol::File>;
+    type HttpTarballRef = super::TarballRef<protocol::HTTP>;
+    type HttpsTarballRef = super::TarballRef<protocol::HTTPS>;
+
+    #[test]
+    fn file_file_roundtrips() {
+        roundtrip::<FileFileRef>("file+file://somewhere/there");
+        roundtrip_to::<FileFileRef>("file://somewhere/there", "file+file://somewhere/there");
```

Comment:
yeah we could check whether the application part is `required` before serializing
I'd like to serialize _more_ rather than losing info, but i'm happy doing whatever nix would do here..